### ISSUE TITLE
BufferGeometry.offsets vs BufferGeometry.drawcalls

### ISSUE THREAD

**Issue by WestLangley** [mrdoob/three.js#5605]
`BufferGeometry` contains the following code for backward compatibility:

	this.drawcalls = [];
	this.offsets = this.drawcalls; // backwards compatibility

Thing is, `BufferGeometry.offsets` gets overwritten in `BufferGeometry.computeOffsets()`:

	this.offsets = offsets;

(`BufferGeometry.computeOffsets()` is called by `CTMLoader`.)

This bug has been hidden, because most of the code accesses `.offsets`. `WireframeHelper`, on the other hand, accesses `.drawcalls`.

---

On a related note, is the 65k addressable limit, which `BufferGeometry.computeOffsets()` addresses, still an issue?

**Comment by iras** [comment]
We've been experiencing problems with this issue too! How's the API-zation going?

**Comment by mrdoob** [comment]
Well, now is even more difficult... https://twitter.com/mrdoob/status/552314480160694273

**Comment by WestLangley** [comment]
Time to revive this.

/ping @dubejf Any ideas?

**Comment by dubejf** [comment]
I don't have much experience with the `drawcalls` / `offsets`. #6803 for rid of all the `.offsets` but I don't have a lot in insight.

> Hmmm... The thing is that I'm still not sure that either is the right name.

drawcalls is a little confusing, it doesn't sound like a data structure. But it's more grep-able than offsets! Plus, "offsets" is starting to be used with instancing.

Thesaurus: chunk, division, region, section, segment, slab, slice, subgeometry, submesh, zone

**Comment by benaadams** [comment]
It does determine the actual drawcalls for the mesh; the offsets arising for when there are more than 65k vertices in a mesh and not being able to be referenced by a 16bit Uint index. (though the ```OES_element_index_uint``` extension removes that limit).

It then means you can now do partial mesh renders, say a couple sliding windows in a object pool (probably contrived).

However, I can see a future change where its also specifies material if that is any help? Say for a MeshFaceMatrial type scenario; or using one mesh with multiple materials.

**Comment by dubejf** [comment]
Yes, to support MeshFaceMaterial, I was assuming that the first drawcall would go with the first material; or maybe add a materialIndex to the drawcalls. But I assume that you already thought of that.

**Comment by benaadams** [comment]
Just thinking naming-wise

Partial renders reference is from there being a few requests for offsets/drawcalls for non-indexed buffers  #6490 #3857 which doesn't have the index limitation; but I'm just guessing at use case.

**Comment by WestLangley** [comment]
`drawcalls` are useful for animating the drawing of lines. I find the name quite descriptive, actually.

**Comment by dubejf** [comment]
@WestLangley what I meant is that, to my ear, a *draw call* is a unit of work on the gpu, whereas what we now know as `drawcalls/offsets` is a decomposition (or a rearranging) of the geometry that can be turned into *draw calls*.

Just like username + password are *credentials*, not *sign-ins*.

**Comment by WestLangley** [comment]
@dubejf Well, yes. But thinking about it a bit more, it seems to me that `drawcalls` (which I feel are reasonably named) and `offsets` should, in fact, be decoupled.

`offsets` is to accommodate a 16-bit limitation in the indexing. `drawcalls` is used to specify the part(s) of the geometry to render. I do not think `offsets` should be manipulated at the API level, but `drawcalls` can be.

That being said, having them the same is not that bad.

**Comment by benaadams** [comment]
@WestLangley Don't think they can be decoupled as altering the offsets generally involves rearranging them and changing the drawcalls

**Comment by WestLangley** [comment]
@benaadams True. I was using the word _decoupled_ in this case to simply mean that `offsets` can be used to break the geometry into groups to accommodate the 65k addressable limit, while `drawcalls` can be used to render a subset of the geometry. In theory, `offsets` and `drawcalls` can be different.

**Comment by jonnenauha** [comment]
I thought that `OES_element_index_uint` was quite universally available and this 65k stuff is not really needed. Is this the case with older browsers/mobile? Or am I talking about a different thing :)

**Comment by benaadams** [comment]
According to http://webglstats.com/ of supported WebGL platforms

Windows 98%
OSX 98%
Linux 96%
ChromeOS 100%
Android Phone 70%
iOS Phone 100%
iOS Tablet 99%
Android Tablet 48%

Other platforms sample sizes to small to be sure

**Comment by mrdoob** [comment]
I'm considering not reimplementing it until someone complains with a good use case...

**Comment by benaadams** [comment]
> I'm considering not reimplementing it until someone complains with a good use case...

Can't even come up with a very good contrived example... 

You can smush a bunch geometries together, including their indexes without renumbering and use the indexOffset to handle the it; but that would result in multiple draws. So, it would make much more sense to renumber the indexes and then it would be a single draw.

As ```OES_element_index_uint``` is now quite widespread; devices that have the 65k limit, you probably don't want to be drawing models >65k anyway?

**Comment by iras** [comment]
Quite widespread? IE11 on Windows 8.1 was complaining about it not long ago.

http://stackoverflow.com/questions/31399639/three-js-r71-not-working-on-ie11-windows-8-1

**Comment by benaadams** [comment]
Depends on DirectX feature level:
9_1 only supports 65k
9_2, 9_3, support 1048575 so wouldn't support OES_element_index_uint
10_0 -> 12.1 support 2^32 – 1 so should support  OES_element_index_uint

For example the original Surface RT doesn't support it; but a 925 Windows Phone on Win10 does.

**Comment by iras** [comment]
Windows 8.1 comes with DirectX 11.2 so according to your table it should already support OES_element_index_uint and 2^32 indexing in IE11, except that it doesn't seem to.

**Comment by benaadams** [comment]
It depends on a bunch of factors including the graphics card in the device. Got to http://webglstats.com/ switch off everything but IE in Personal Computer; scroll down till you get to OES_element_index_uint and you'll see it appeared in August 2014 with 31% support and had 77% by September 2014; and currently is in July 2015 has 93.9% support.

The SO question you linked to was written last year in July 2014 before IE had much support for it.

**Comment by iras** [comment]
The SO question was asked a month ago, July 14th, also mentions Three.js r71 which is the latest.

Thank you for the webGL stats breakdown anyway. I'll do some more tests to see whether the situation has somehow improved with IE11 in Windows 8.1 in the past month.

**Comment by benaadams** [comment]
> The SO question was asked a month ago, July 14th

Lol - my mistake; July 14 reads to me as a month and a year :)

**Comment by mrdoob** [comment]
> You can smush a bunch geometries together, including their indexes without renumbering and use the indexOffset to handle the it; but that would result in multiple draws. So, it would make much more sense to renumber the indexes and then it would be a single draw.

It's always multiple draw calls. I guess the main benefit is that we don't have to re-upload the position/normal/color buffer. However, I don't think the code ever did that. I think we have to do more refactoring around all this for it to actually make sense.

**Comment by bbeausej** [comment]
I wrote the computeOffsets / reorderBuffers code. My use case was loading meshes of several million polygons using the CTMLoader and BufferGeometry. Mind you this was a while ago and might not apply today anymore...

I would hit the 65k limit in Chrome at the time. The CTMLoader had prototype code to try to deal with this but failed miserably and would generate several thousand offsets/draw calls for my meshes.

The particular functions I added then for BufferGeometry allowed the CTMLoader to simply push those meshes to the BufferGeometry and correct the offsets for draw. This allowed for a very streamlined loading process for large meshes using CTM.

-b
